File: src/main/cpp/lib/Invoice.cpp
/*******************************************************************************
 *
 * Copyright (c) {2003-2019} Murex S.A.S. and its affiliates.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 *******************************************************************************/

#include "purchase/Invoice.h"
#include "purchase/PurchasedBook.h"
#include "IdGenerator.h"
#include "finance/TaxRule.h"

purchase::Invoice::Invoice(const std::string& clientName, const domain::country::Country& country)
: Invoice(id_generator::nextId(), clientName, country)
{
}

purchase::Invoice::Invoice(int id, const std::string& clientName, const domain::country::Country& country)
	: id_(id)
    , clientName_(clientName)
    , country_(country)
{
}

void purchase::Invoice::addPurchasedBooks(std::vector<std::shared_ptr<const PurchasedBook>> books)
{
	purchasedBooks_.insert(purchasedBooks_.end(), books.begin(), books.end());
}

void purchase::Invoice::addPurchasedBook(const std::shared_ptr<const PurchasedBook>& book)
{
	purchasedBooks_.push_back(book);
}

double purchase::Invoice::computeTotalAmount() const
{
	double sum = 0.0;
	for (const auto purchasedBook : purchasedBooks_)
	{
		double totalPrice = purchasedBook->getTotalPrice() * finance::getApplicableRate(country_, *purchasedBook->getBook());
		sum += totalPrice;
	}
	return sum;
}

File: src/main/csharp/Application/Purchase/Invoice.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using Application.Domain.Country;

namespace Application.Purchase
{
    public sealed class Invoice
    {
        public int Id { get; }
        public string ClientName { get; }
        public Country Country { get; }
        public List<PurchasedBook> PurchasedBooks { get; }


        public Invoice(string clientName, Country country) 
            : this(IdGenerator.NextId(), clientName, country)
        {
        }

        public Invoice(int id, string clientName, Country country)
        {
            Id = id;
            ClientName = clientName;
            Country = country;
            PurchasedBooks = new List<PurchasedBook>();
        }

        public void AddPurchasedBooks(List<PurchasedBook> books)
        {
            PurchasedBooks.AddRange(books);
        }

        public double ComputeTotalAmount()
        {
            var totalAmount = 0.0;
            totalAmount = PurchasedBooks.Sum(book => book.TotalPrice);
            return totalAmount;
        }

        public override string ToString() 
            => $"Invoice [ {nameof(Id)}: '{Id}'" +
               $", {nameof(ClientName)}: '{ClientName}'" +
               $", {nameof(Country)}: '{Country}'" +
               $", {nameof(PurchasedBooks)}: '{PurchasedBooks}' ]";

        private bool Equals(Invoice other)
        {
            return Id == other.Id
                   && ClientName == other.ClientName
                   && Equals(Country, other.Country) 
                   && Equals(PurchasedBooks, other.PurchasedBooks);
        }

        public override bool Equals(object obj)
        {
            return ReferenceEquals(this, obj) || obj is Invoice other && Equals(other);
        }

        public override int GetHashCode() => HashCode.Combine(Id, ClientName, Country, PurchasedBooks);
    }
}